tools/sof_perf_analyzer.py

```diff
@@ -182,11 +179,19 @@ def process_kmsg_file():
     [  334.818442] kernel: snd_sof:sof_ipc4_log_header: sof-audio-pci-intel-mtl 0000:00:1f.3:
     ipc tx      : 0x40000004|0x15: MOD_INIT_INSTANCE [data size: 84]
 
+    However, for some cases, the line after "Creating" does not contain widget ID information,
```

Comment:
In which cases? Is that not a kernel bug that this PR would then hide?

cc: @plbossart , @ranj063

Reply:
@btian1 what exactly is the problem statement? You are suggesting a solution without giving any context or alternate suggestion. That's not very helpful, I have no idea what this patch does.

Reply:
@plbossart , please check: https://github.com/thesofproject/sof-test/issues/1108
there is a possibility of log missed for kernel log output, in this case, this patch try to parse bind related info to get widget ID, widget name and so on.

@marc-hb , I checked kernel code, there should have no bug in kernel, just kernel log missing with a possibility:
https://github.com/thesofproject/linux/blob/topic/sof-dev/sound/soc/sof/ipc4.c#L222
@plbossart , please double confirm.

Reply:
I will double-confirm that I don't understand the problem statement and why you are not trying to fix the issue (if it's real) at the source by updating the kernel logs.

Reply:
Thanks, @plbossart , let me use logs to describe:
expected logs:
[61230.667725] kernel: snd_sof:sof_ipc4_widget_setup: sof-audio-pci-intel-mtl 0000:00:1f.3: Create widget mixin.20.1 instance 0 - pipe 20 - core 0
[61230.667728] kernel: snd_sof:sof_ipc4_log_header: sof-audio-pci-intel-mtl 0000:00:1f.3: ipc tx      : 0x40000002|0xa: MOD_INIT_INSTANCE [data size: 40]

we can get name(mixin.20.1) and its module ID: 2(0x40000002).

unexpected logs(may not have errror in kernel):

[  262.012482] kernel: snd_sof:sof_ipc4_widget_setup: sof-audio-pci-intel-mtl 0000:00:1f.3: Create widget alh-copier.SDW0-Capture.0 instance 8 - pipe 11 - core 0
[  262.012486] kernel: snd_sof_intel_hda_common:hda_dsp_stream_trigger: sof-audio-pci-intel-mtl 0000:00:1f.3: FW Poll Status: reg[0xa0]=0x2024001e successful

[  261.979993] kernel: snd_sof:sof_ipc4_widget_setup: sof-audio-pci-intel-mtl 0000:00:1f.3: Create widget pipeline.31 instance 4 - pipe 31 - core 0
[  261.980114] kernel: snd_sof:sof_ipc4_log_header: sof-audio-pci-intel-mtl 0000:00:1f.3: ipc tx reply: 0x33000000|0x1: GLB_SET_PIPELINE_STATE

[61230.948775] kernel: snd_sof:sof_ipc4_widget_setup: sof-audio-pci-intel-mtl 0000:00:1f.3: Create widget gain.15.1 instance 4 - pipe 15 - core 0
[61230.948777] kernel: snd_sof:sof_pcm_trigger: sof-audio-pci-intel-mtl 0000:00:1f.3: pcm: trigger stream 0 dir 0 cmd 1

so we get those info from bind related logs:
[61230.670335] kernel: snd_sof:sof_ipc4_route_setup: sof-audio-pci-intel-mtl 0000:00:1f.3: bind host-copier.2.playback:0 -> gain.20.1:0
[61230.670338] kernel: snd_sof:sof_ipc4_log_header: sof-audio-pci-intel-mtl 0000:00:1f.3: ipc tx      : 0x45000004|0x6: MOD_BIND

get name(host-copier.2.playback), ID 4(0x45000004).

this is this patch's purpose.

Reply:
I am unable to follow what's missing or not right in the logs above.

It's rather disappointing that after multiple asks I still get this sort of raw dump of information without clarity on the problem statement. This is starting to be a waste of everyone's time if I am honest.

Reply:
@plbossart I chatted with @btian1. The purpose is to connect the widget name and the IPC primary data. Like `mixin.20.1` <-> `0x40000002`.
The existing code get the information from dmesg log when it looks for `Create widget` and parse the log in 2 lines. It works in most cases, but as you know, the kernel is multi thread and there is no guarantee that the `ipc tx :` log will be right after the `Create widget` log. And the existing code will return error in that case.
The commit will use another key work `bind` to get the IPC primary data. @btian1 @aiChaoSONG Please correct me if I was wrong.

It should be easy to "solve" the issue if we can add msg->primary to 
```
dev_dbg(sdev->dev, "Create widget %s instance %d - pipe %d - core %d\n",
                swidget->widget->name, swidget->instance_id, swidget->pipeline_id, swidget->core);
```
Note that I don't mean that we need to fix it from kernel. But trying to explain the issue they are trying to fix.

Reply:
> It should be easy to "solve" the issue if we can add msg->primary to

Sounds great to me! @plbossart ?

Reply:
The IPC is serialized so the order will not vary. There might be other things in between IPCs but they will be by construction based on different/irrelevant information.

So yes if we can change the logs there's no issue. This isn't an ABI, we can tune the logs to the needs of profiling tools.

Reply:
Sorry for jumping in a bit late. This is not an error in the kernel ogs nor an unexpected sequence of logs. When widgets in a pipeline are set up, the first widget in the pipeline sets up the pipeline before the widget itself gets set up. So please do not assume that the log ```dev_dbg(sdev->dev, "Create widget %s instance %d - pipe %d - core %d\n",...```. If you do , please fix it instead of assuming incorrect sequences.

Reply:
not able to follow your answer @ranj063, there's something missing in the sentence "do not assume that the log"

Reply:
The swidget is a Linux SOF concept and it represents all sorts of 'Sof WIDGETs'.
The pipeline is also a swidget and that is not  a module, it has no 'module ID'

Reply:
My proposal is that the kernel should print two version for swidget creation.
For 'modules': `Create widget for module: %s, module id: %u, instance id: %u, pipeline: %u, core: %u`
For 'pipelines': `Create widget for pipeline: %s, id: %u, instance id: %u, core: %u`

The pipeline_id is a Linux driver concept, it is good to have it out, but not sure what would be the best way.